Segment anything pipeline image

pschroedl opened the pull request:
**What does this pull request do? Explain your changes. (required)**

Adds support for the ai-worker pipeline which implements Segment Anything 2 from facebook/meta
 (https://github.com/facebookresearch/segment-anything-2)

Corresponding AI-Worker PR:
https://github.com/livepeer/ai-worker/pull/185

**Specific updates (required)**

~~* Adds a mapping of pipelines to a tag corresponding to the pipeline-specific docker image~~
Mapping of pipelines to a tag corresponding to the pipeline-specific docker image has been pushed down to the ai-worker
* Launches a container with the correct image for pipelines specified warm == True in aiModels.json
* Pricing for semgent-anything-2 requests

**How did you test each of these updates (required)**

E2E testing gateway + orchestrator via curl and jupyter notebook

Example usage: 
https://colab.research.google.com/drive/12TC3NW7BpO77WfKHQyok18ebOtbT9F4-?usp=sharing

**Does this pull request close any open issues?**



**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Read the [contribution guide](./CONTRIBUTING.md)
- [x] `make` runs successfully
- [ ] All tests in `./test.sh` pass
- [ ] README and other documentation updated
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

pschroedl committed:
use pipeline to image mapping, update ai-worker interface for segment anything

pschroedl committed:
fix comma typo, remove incorrect latency calculation and return type

pschroedl committed:
NewSegmentAnything2Writer -> worker.worker.NewSegmentAnything2MultipartWriter

pschroedl committed:
fix missing process method and capability naming

pschroedl committed:
add rpc handler

eliteprox committed:
add pricing and latency calculations

pschroedl committed:
Merge pull request #1 from eliteprox/segment-anything-pipeline-image

add pricing and latency calculations

Commit:
move pipeline -> image mapping to core/ai

rickstaa commented on core/capabilities.go:
```diff
@@ -114,6 +115,7 @@ var CapabilityNameLookup = map[Capability]string{
 	Capability_ImageToVideo:               "Image to video",
 	Capability_Upscale:                    "Upscale",
 	Capability_AudioToText:                "Audio to text",
```
@pschroedl For now, I think it's perfectly fine to add this as a hardcoded capability 👍🏻. In the future, we might want to explore making this mechanism more generic, but let's save that challenge for another time 😅.

rickstaa commented on cmd/livepeer/starter/starter.go:
```diff
@@ -1227,10 +1227,27 @@ func StartLivepeer(ctx context.Context, cfg LivepeerConfig) {
+						}
+
+						// Setup the AI worker with the selected Docker image
+						n.AIWorker, err = worker.NewWorker(imageToUse, gpus, modelsDir)
```
@pschroedll Unfortunately, this logic won't work when starting multiple containers. Here's what I suggest we do instead:

- Rename the [containerImageID](https://github.com/pschroedl/ai-worker/blob/50fcc14304dc4b6d7ea8a7cf055b32a35d9d5768/worker/docker.go#L43) to `containerImageIDDefault`.
- Add an optional `containerImageID` parameter to the [function](https://vscode.dev/github/pschroedl/ai-worker/blob/segment_anything_2_pipeline_image/worker/docker.go#L157) and set it in the container config when present.
- Apply the same logic to [docker.go/DockerManager/Borrow](https://github.com/pschroedl/ai-worker/blob/50fcc14304dc4b6d7ea8a7cf055b32a35d9d5768/worker/docker.go#L110).
- Apply the same logic to [docker.go/DockerManager/Warm](https://github.com/pschroedl/ai-worker/blob/50fcc14304dc4b6d7ea8a7cf055b32a35d9d5768/worker/docker.go#L79).
- This will allow us to simplify the logic in the starter as well.

Let me know your thoughts!

pschroedl replied:
Sounds like a good approach to me - thanks @rickstaa !

pschroedl replied:
After testing this approach, I realized that a better approach might be to delegate custom image mapping down to the AI-Worker. 

One motivation is that Borrow is also called directly from ai-worker itself, leading to a state where the the custom containerImageID name:tag was unavailable.

rickstaa commented on cmd/livepeer/starter/starter.go:
```diff
@@ -1227,10 +1227,27 @@ func StartLivepeer(ctx context.Context, cfg LivepeerConfig) {
+					// If config.Warm is true, setup the AI worker with the selected Docker image
+					if config.Warm && config.URL == "" {
+						// Determine which image to use based on the pipeline
+						imageToUse := *cfg.AIRunnerImage // Default to the image from config
```
@pschroedl can we replace this with a helperfunction like `getContainerImageID`?

pschroedl replied:
OK, SGTM.

rickstaa reviewed (changes requested):
@pschroedll Please check out my requested changed. If those are dealt with it is ready for release.

rickstaa committed:
refactor(ai): improve comments

This commit applies some small comment changes.

rickstaa committed:
fix(ai): fix broken Gateway SAM2 latency score

This commit fixes the orch latency score at the Gateway for the SAM2 pipeline
which was commented out in a previous commit.

rickstaa commented on server/ai_process.go:
```diff
@@ -586,6 +587,102 @@ func submitUpscale(ctx context.Context, params aiRequestParams, sess *AISession,
+
+	// start := time.Now()
+	resp, err := client.SegmentAnything2WithBodyWithResponse(ctx, mw.FormDataContentType(), &buf, setHeaders)
+	// took := time.Since(start)
```
@pschroedl I enabled this again in https://github.com/livepeer/go-livepeer/pull/3131/commits/6d3ce7315f5485852f233bc8d34719d45cb25c16. Was something wrong with the calculation?

pschroedl replied:
I had only intended to disable the timer until the calculation method had been updated - good catch!

pschroedl committed:
wip container image handling

pschroedl committed:
push image selection to worker

pschroedl committed:
add back config derived default iage

rickstaa reviewed (approved):
LGTM 🚀 !

rickstaa commented:
@pschroedl, only thing left is updating the worker version before merge.

pschroedl committed:
bump ai-worker version in go.mod

pschroedl committed:
add segmentanything2 to orch stub methods

rickstaa commented:
@pschroedl although some tests pass these are the same as the ones on the `ai-video` main branch. We still need to fix those and write new tests for the AI stuff. Let's prioritize this after the public launch.
